lib/sync/mongodbQueue.js

```diff
@@ -55,11 +57,75 @@ MongodbQueue.prototype.create = function(cb) {
-    self.queue.createIndexes(function(err) {
+    var collection = self.mongodb.collection(self.queueName);
+    var indexName = 'deleted_1';
+    self.lock.acquire(self.lockName, self.lockTimeout, function(err, lockCode){
       if (err) {
+        debugError('[%s] failed to acquire lock %s due to error : %s', self.queueName, self.lockName, err);
         return cb(err);
       }
-      return cb(null, self.queue);
+      if (lockCode) {
+        debug('[%s] lock %s acquired. Continue.', self.queueName, self.lockName);
+        async.waterfall([
+          function createIndexes(callback) {
+            debug('[%s] creating queue indexes', self.queueName);
+            self.queue.createIndexes(callback);
+          },
+          
+          function listIndexes(createdIndex, callback) {
+            debug('[%s] list existing indexes', self.queueName);
+            collection.indexInformation({full: true}, callback);
+          },
+          function checkIndexTTL(indexInfo, callback) {
+            debug('[%s] found queue indexInfo : %j', self.queueName, indexInfo);
+            var existingIndex = _.findWhere(indexInfo, {name: indexName});
+            var needDrop = false;
+            var needCreate = false;
+
+            if (!existingIndex) {
+              needCreate = true;
+            } else if (existingIndex.expireAfterSeconds !== self.queueTTL) {
+              needDrop = true;
+              needCreate = true;
+            }
+            return callback(null, needDrop, needCreate);
+          },
+          function dropTTLIndex(needDrop, needCreate, callback) {
+            if (needDrop) {
+              debug('[%s] dropping ttl index: %s', self.queueName, indexName);
+              collection.dropIndex(indexName, function(err){
+                return callback(err, needCreate);
+              });
+            } else {
+              debug('[%s] skip dropping ttl index', self.queueName);
+              return callback(null, needCreate);
+            }
+          },
+          function createTTLIndex(needCreate, callback) {
+            if (needCreate) {
+              debug('[%s] creating ttl index: %s', self.queueName, indexName);
+              collection.createIndex({'deleted': 1}, {'expireAfterSeconds': self.queueTTL, 'backgroud': true}, callback);
+            } else {
+              debug('[%s] skip creating ttl index', self.queueName);
+              return callback();
+            }
+          }
+        ], function(err){
+          if (err) {
+            debugError('[%s] failed to create queue index due to error: %s', self.queueName, err);
+            return cb(err);
+          }
+          self.lock.release(self.lockName, lockCode, function(releaseErr){
+            if (releaseErr) {
+              debugError('[%s] failed to release lock due to error: %s', self.queueName, releaseErr);
+            }
+          });
+          return cb(null, self.queue);
+        });
+      } else {
```

Comment:
from a code readability point of view, it might be better to take case of this else case first as its kind of hidden at the bottom here